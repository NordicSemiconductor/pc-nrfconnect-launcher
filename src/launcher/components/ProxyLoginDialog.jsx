/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import InputGroup from 'react-bootstrap/InputGroup';
import Modal from 'react-bootstrap/Modal';
import { bool, func, string } from 'prop-types';

class ProxyLoginDialog extends React.Component {
    constructor() {
        super();
        this.state = {
            password: '',
        };
        this.onUserChanged = this.onUserChanged.bind(this);
        this.onPasswordChanged = this.onPasswordChanged.bind(this);
        this.onSubmitClicked = this.onSubmitClicked.bind(this);
        this.onKeyPress = this.onKeyPress.bind(this);
    }

    onUserChanged(event) {
        const { onUsernameChanged } = this.props;
        onUsernameChanged(event.target.value);
    }

    onPasswordChanged(event) {
        this.setState({
            password: event.target.value,
        });
    }

    onSubmitClicked() {
        const { onSubmit, username } = this.props;
        const { password } = this.state;
        onSubmit(username, password);
        this.setState({
            password: '',
        });
    }

    onKeyPress(event) {
        if (event.key === 'Enter' && this.isValidInput()) {
            this.onSubmitClicked();
        }
    }

    isValidInput() {
        const { username } = this.props;
        const { password } = this.state;
        return username !== '' && password !== '';
    }

    render() {
        const { isVisible, message, onCancel, username } = this.props;
        const { password } = this.state;
        return (
            <Modal show={isVisible} backdrop>
                <Modal.Header closeButton={false}>
                    <Modal.Title>Proxy authentication required</Modal.Title>
                </Modal.Header>
                <Modal.Body>
                    <p>{message}</p>
                    <Form.Group controlId="username">
                        <Form.Label>Username:</Form.Label>
                        <InputGroup>
                            <Form.Control
                                autoFocus
                                value={username}
                                onChange={this.onUserChanged}
                                onKeyPress={this.onKeyPress}
                            />
                        </InputGroup>
                    </Form.Group>
                    <Form.Group controlId="password">
                        <Form.Label>Password:</Form.Label>
                        <InputGroup>
                            <Form.Control
                                value={password}
                                type="password"
                                onChange={this.onPasswordChanged}
                                onKeyPress={this.onKeyPress}
                            />
                        </InputGroup>
                    </Form.Group>
                </Modal.Body>
                <Modal.Footer>
                    <Button onClick={onCancel}>Cancel</Button>
                    <Button
                        onClick={this.onSubmitClicked}
                        disabled={!this.isValidInput()}
                    >
                        Login
                    </Button>
                </Modal.Footer>
            </Modal>
        );
    }
}

ProxyLoginDialog.propTypes = {
    isVisible: bool.isRequired,
    message: string.isRequired,
    username: string.isRequired,
    onUsernameChanged: func.isRequired,
    onCancel: func.isRequired,
    onSubmit: func.isRequired,
};

export default ProxyLoginDialog;
