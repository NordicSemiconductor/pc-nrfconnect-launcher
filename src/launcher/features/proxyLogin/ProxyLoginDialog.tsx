/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React, { useCallback, useMemo, useState } from 'react';
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import InputGroup from 'react-bootstrap/InputGroup';
import Modal from 'react-bootstrap/Modal';

import { answerProxyLoginRequest } from '../../../ipc/proxyLogin';
import { useLauncherDispatch, useLauncherSelector } from '../../util/hooks';
import {
    changeUserName,
    getProxyLoginRequest,
    loginCancelledByUser,
    loginRequestSent,
} from './proxyLoginSlice';

export default () => {
    const dispatch = useLauncherDispatch();
    const [password, setPassword] = useState('');

    const { isVisible, username, host, requestId } =
        useLauncherSelector(getProxyLoginRequest);

    const cancel = useCallback(() => {
        answerProxyLoginRequest(requestId!); // eslint-disable-line @typescript-eslint/no-non-null-assertion
        dispatch(loginCancelledByUser());
    }, [dispatch, requestId]);

    const submit = useCallback(() => {
        answerProxyLoginRequest(requestId!, username, password); // eslint-disable-line @typescript-eslint/no-non-null-assertion
        dispatch(loginRequestSent());
        setPassword('');
    }, [dispatch, password, requestId, username]);

    const inputIsValid = useMemo(
        () => username !== '' && password !== '',
        [password, username]
    );

    const submitOnEnter: React.KeyboardEventHandler = useCallback(
        event => {
            if (event.key === 'Enter' && inputIsValid) {
                submit();
            }
        },
        [inputIsValid, submit]
    );

    return (
        <Modal show={isVisible} backdrop>
            <Modal.Header closeButton={false}>
                <Modal.Title>Proxy authentication required</Modal.Title>
            </Modal.Header>
            <Modal.Body>
                <p>
                    The proxy server {host} requires authentication. Please
                    enter username and password
                </p>
                <Form.Group controlId="username">
                    <Form.Label>Username:</Form.Label>
                    <InputGroup>
                        <Form.Control
                            autoFocus
                            value={username}
                            onChange={event =>
                                dispatch(changeUserName(event.target.value))
                            }
                            onKeyPress={submitOnEnter}
                        />
                    </InputGroup>
                </Form.Group>
                <Form.Group controlId="password">
                    <Form.Label>Password:</Form.Label>
                    <InputGroup>
                        <Form.Control
                            value={password}
                            type="password"
                            onChange={event => setPassword(event.target.value)}
                            onKeyPress={submitOnEnter}
                        />
                    </InputGroup>
                </Form.Group>
            </Modal.Body>
            <Modal.Footer>
                <Button onClick={cancel}>Cancel</Button>
                <Button onClick={submit} disabled={!inputIsValid}>
                    Login
                </Button>
            </Modal.Footer>
        </Modal>
    );
};
