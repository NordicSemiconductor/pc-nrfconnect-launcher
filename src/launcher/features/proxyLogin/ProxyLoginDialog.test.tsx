/*
 * Copyright (c) 2022 Nordic Semiconductor ASA
 *
 * SPDX-License-Identifier: LicenseRef-Nordic-4-Clause
 */

import React from 'react';

import render from '../../../testrenderer';
import ProxyLoginDialog from './ProxyLoginDialog';
import {
    changeUserName,
    loginCancelledByUser,
    loginRequestedByServer,
    loginRequestSent,
} from './proxyLoginSlice';

const requestId = 'test ID';
const authInfo = {
    isProxy: true,
    scheme: 'https',
    host: 'test host',
    port: 4711,
    realm: 'test realm',
};

describe('ProxyLoginDialog', () => {
    it('is initially invisible', () => {
        expect(render(<ProxyLoginDialog />).baseElement).toMatchSnapshot();
    });

    it('is displayed after a login is requested', () => {
        expect(
            render(<ProxyLoginDialog />, [
                loginRequestedByServer({ requestId, authInfo }),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('is updated if users enter a username', () => {
        expect(
            render(<ProxyLoginDialog />, [
                loginRequestedByServer({ requestId, authInfo }),
                changeUserName('new username'),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('is invisible after users cancel the login', () => {
        expect(
            render(<ProxyLoginDialog />, [
                loginRequestedByServer({ requestId, authInfo }),
                loginCancelledByUser(),
            ]).baseElement
        ).toMatchSnapshot();
    });

    it('is invisible after users enter username and password', () => {
        expect(
            render(<ProxyLoginDialog />, [
                loginRequestedByServer({ requestId, authInfo }),
                loginRequestSent('the username'),
            ]).baseElement
        ).toMatchSnapshot();
    });
});
