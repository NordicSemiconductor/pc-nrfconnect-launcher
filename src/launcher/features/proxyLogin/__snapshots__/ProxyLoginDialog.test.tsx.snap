// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProxyLoginDialog is displayed after a login is requested 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
          >
            Proxy authentication required
          </div>
        </div>
        <div
          class="modal-body"
        >
          <p>
            The proxy server 
            test host (realm: test realm)
             requires authentication. Please enter username and password
          </p>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="username"
            >
              Username:
            </label>
            <div
              class="input-group"
            >
              <input
                class="form-control focus-visible"
                data-focus-visible-added=""
                id="username"
                value=""
              />
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="password"
            >
              Password:
            </label>
            <div
              class="input-group"
            >
              <input
                class="form-control"
                id="password"
                type="password"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Cancel
          </button>
          <button
            class="btn btn-primary"
            disabled=""
            type="button"
          >
            Login
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ProxyLoginDialog is initially invisible 1`] = `
<body>
  <div />
</body>
`;

exports[`ProxyLoginDialog is invisible after users cancel the login 1`] = `
<body
  class=""
  style="padding-right: 0px;"
>
  <div />
</body>
`;

exports[`ProxyLoginDialog is invisible after users enter username and password 1`] = `
<body
  class=""
  style="padding-right: 0px;"
>
  <div />
</body>
`;

exports[`ProxyLoginDialog is updated if users enter a username 1`] = `
<body
  class="modal-open"
  style="padding-right: 0px;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="fade modal-backdrop show"
  />
  <div
    aria-modal="true"
    class="fade modal show"
    role="dialog"
    style="display: block;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <div
            class="modal-title h4"
          >
            Proxy authentication required
          </div>
        </div>
        <div
          class="modal-body"
        >
          <p>
            The proxy server 
            test host (realm: test realm)
             requires authentication. Please enter username and password
          </p>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="username"
            >
              Username:
            </label>
            <div
              class="input-group"
            >
              <input
                class="form-control focus-visible"
                data-focus-visible-added=""
                id="username"
                value="new username"
              />
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="password"
            >
              Password:
            </label>
            <div
              class="input-group"
            >
              <input
                class="form-control"
                id="password"
                type="password"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Cancel
          </button>
          <button
            class="btn btn-primary"
            disabled=""
            type="button"
          >
            Login
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
